kubernetes/updater/pkg/maintenance/trigger.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package maintenance
+
+import (
+	"context"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Trigger is evaluated to decide whether a maintenance can happen or not.
+// Maintenances can happen because of multiple reasons like:
+// - attempt to recover from a broken state
+// - we are in a maintenance window
+// - emergency security patch
+// Each Trigger has a Name() for logging purposes and a Default() method
+// returning whether the trigger should fail open or close in case of error.
+type Trigger interface {
+	Name() string
+	CanStart(ctx context.Context, object v1.Object) (bool, error)
+	Default() bool
+}
+
+// Triggers is a list of Trigger. Triggers are OR-ed: any trigger firing in the
+// list will cause the maintenance to be triggered.
+type Triggers []Trigger
+
+func (t Triggers) AtLeastOne(ctx context.Context, object v1.Object) bool {
+	log := ctrllog.FromContext(ctx).V(1)
+	for _, trigger := range t {
+		start, err := trigger.CanStart(ctx, object)
+		if err != nil {
+			start = trigger.Default()
+			log.Error(err, "trigger failed to evaluate, using its default value", "trigger", trigger.Name(), "defaultValue", start)
+		} else {
+			log.Info("trigger evaluated", "trigger", trigger.Name(), "result", start)
```

Comment:
This feels more like debug.

Reply:
This is not `logrus` but `logr` (because this is the login interface the controller runtime chose). Debug logs are done by increasing the verbosity level and calling `Info()`. The verbosity is already set to DEBUG for the whole function (this happens on the line getting the logger from the context).